### PR TITLE
Implemented registration

### PR DESCRIPTION
**Pull request by vladNed**
Implemented registration flow and added the following endpoint (in order) for completing a registration process:

For user:
- `startPoolRegistration` - initializes the registration and pays the fee
- `readyPoolRegistration` - adds the initial liquidity and makes the registration ready for review

For admin:
- `apprivePoolRegistration`/`rejectPoolRegistration` - endpoints to either deploy the child contract once approved or reject and send everything back to the user.
- `completePoolRegistration` - once the admin has issued the LP token and set necessary roles this endpoint is called to send the initial liquidity added by the user and complete the registration process.

Possible improvements: 
- [x] Add a view for the user to grab all his approved and deployed pool
- [x] Add step that can reject if NEW state
- [x] Transitions for all states in admin
- [ ] Events for all transition changes

### CODE REVIEW

File: router/src/register.rs [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+    #[payable("EGLD")]
+    #[endpoint(startPoolRegistration)]
+    fn start_pool_registration(
+        &self,
+        token_id: TokenIdentifier,
+        collection_id: TokenIdentifier,
+        delta: BigUint,
+        spot_price: BigUint,
+        bonding_curve: BondingCurve,
+        opt_fee_percents: OptionalValue<MultiValue2<u64, u64>>
```
Comment (alexjilavu) [inline router/src/register.rs:113]: Will this be of any use? We shall not let the user set his own fee percent amounts, so I would remove this.

File: router/src/register.rs [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+        self.registration_queue().insert(pair_key, registration);
+    }
+
+    #[only_owner]
+    #[endpoint(rejectPoolRegistration)]
+    fn reject_pool_registration(&self, token_id: TokenIdentifier, collection_id: TokenIdentifier) {
+        let pair_key = (token_id.clone(), collection_id.clone());
+        let registration = match self.registration_queue().get(&pair_key) {
+            Some(value) => {
+                require!(value.state == RegistrationState::READY, "This registration is not completed");
```
Comment (alexjilavu) [inline router/src/register.rs:177]: I would also allow the state NEW to be rejected. What if the user only pays the fee and does not send the initial liquidity or an error occurs?

File: router/src/register.rs [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+                value
+            },
+            None => sc_panic!("No registration was made for this pool")
+        };
+
+        match registration.payments {
+            Some(value) => self.send_multiple_tokens_if_not_zero(&registration.pool_owner, &value),
+            None => sc_panic!("Could not find any registration payments")
```
Comment (alexjilavu) [inline router/src/register.rs:185]: If we also allow RegistrationState::NEW to be rejected this panic must be removed

File: router/src/register.rs [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+    #[callback]
+    fn lp_token_issue_callback(
+        &self,
+        caller: &ManagedAddress,
+        address: &ManagedAddress,
+        #[call_result] result: ManagedAsyncCallResult<()>,
+    ) {
+        let (token_id, returned_tokens) = self.call_value().egld_or_single_fungible_esdt();
+        match result {
+            ManagedAsyncCallResult::Ok(()) => {
```
Comment (alexjilavu) [inline router/src/register.rs:344]: Maybe also transition the registration to a new state like LP_TOKEN_ISSUED here for better traceability?

File: router/src/register.rs [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+    #[only_owner]
+    #[endpoint(setLocalRoles)]
+    fn set_local_roles(&self, pair_address: ManagedAddress) {
+        // require!(self.state().get(), "Not active");
+        // TODO
+        self.check_is_pair_sc(&pair_address);
+
+        let token_pair = self.get_pair_by_address(&pair_address);
+        let mut registration = self.get_registration(&token_pair);
+        require!(registration.state == RegistrationState::APPROVED, "This pool registration was not approved yet");
```
Comment (alexjilavu) [inline router/src/register.rs:253]: registration.state == RegistrationState::LP_TOKEN_ISSUED

**Review by alexjilavu — CHANGES_REQUESTED** [review]

File: router/src/register.rs (hunk 1 above)
Comment (vladNed) [reply router/src/register.rs]: removed the optional fees from the endpoint

File: router/src/register.rs (hunk 2 above)
Comment (vladNed) [reply router/src/register.rs]: made a new status for this

**Review by alexjilavu — APPROVED** [review]

**Pull request merged**